fix(docs-infra): several small a11y improvements (including a "skip to content" link)

gkalpak opened the pull request:
This PR contains several small accessibility improvements for angular.io (mainly around navigating the page using a keyboard).
See the individual commits for more details.

---
The most impactful change is the addition of a "skip to content" link (last commit):

**fix(docs-infra): add a "skip to content" link for better accessibility**:

Add a link at the top of the page that allows skipping directly to the main content area. This allows people using assistive technologies to skip navigation-related parts of the page which are presented before the main content.

Related resources:
- [www.w3.org](https://www.w3.org/TR/WCAG20-TECHS/G1.html)
- [accessibility.oit.ncsu.edu](https://accessibility.oit.ncsu.edu/it-accessibility-at-nc-state/developers/accessibility-handbook/mouse-and-keyboard-events/skip-to-main-content)
- [css-tricks.com](https://css-tricks.com/how-to-create-a-skip-to-content-link)

##
Example implementations on other websites:

- **[accessibility.oit.ncsu.edu](https://accessibility.oit.ncsu.edu/)**:
  ![accessibility.oit.ncsu.edu](https://user-images.githubusercontent.com/8604205/133285834-db1314e6-ff71-4862-8f80-c8a9f69083d9.png)

- **[smile.amazon.com](https://smile.amazon.com/)**:
  ![smile.amazon.com](https://user-images.githubusercontent.com/8604205/133285854-1163e733-8de3-4244-bc23-5cd5a943001d.png)

- **[chase.com](https://chase.com/)**:
  ![chase.com](https://user-images.githubusercontent.com/8604205/133293396-ade9d95f-56ce-4cdc-8926-b4cd2dc6c3a5.png)

- **[css-tricks.com](https://css-tricks.com/)**:
  ![css-tricks.com](https://user-images.githubusercontent.com/8604205/133285845-702045ee-088a-40f1-bb26-21adebb02505.png)

- **[github.com/angular/angular](https://github.com/angular/angular)**:
  ![github.com/angular/angular](https://user-images.githubusercontent.com/8604205/133921259-9fd759b9-f862-47c3-b069-643912df17a1.png)

- **[news.sky.com](https://news.sky.com/)**:
  ![news.sky.com](https://user-images.githubusercontent.com/8604205/133285863-89946096-5353-4b2b-ab69-eb748bdcfe06.png)

- **[youtube.com](https://youtube.com/)**:
  ![youtube.com](https://user-images.githubusercontent.com/8604205/133921261-a68a12c1-1aea-4ad0-8457-b4eaf8ed3497.png)

##
Angular.io implementation:

- **[angular.io](https://angular.io/)**:
  ![angular.io](https://user-images.githubusercontent.com/8604205/133409634-0054625c-d14b-4ebe-bb4c-142cfc524d98.png)

---
Fixes #42936.

mary-poppins commented:
You can preview 8554efa at https://pr43460-8554efa.ngbuilds.io/.

petebacondarwin commented on aio/src/app/custom-elements/contributor/contributor.component.ts:
```diff
@@ -28,7 +28,7 @@ import { CONTENT_URL_PREFIX } from 'app/documents/document.service';
         </div>
 
-        <div class="card-back" *ngIf="person.isFlipped" (click)="flipCard(person)" (keyup.enter)="flipCard(person)">
+        <div class="card-back" *ngIf="person.isFlipped" (click)="flipCard(person)" (keyup.enter)="flipCard(person)" tabindex="0">
```
I notice that once the person is flipped the bio is not actually focussed. So you then need to tab one more time before you can interact with the bio (and flip it back).

Perhaps when we flip we could consider setting focus on the bio when it is shown?
Or perhaps just remove the `card-front` element (using another `*ngIf` which I think would move focus to the bio...

gkalpak replied:
Yeah, I noticed that too. Unfortunately, I could not come up with a good way to fix that (without complicating the component with view queries etc., which I felt was not worth it). I also tried removing `card-front` with an `*ngIf`, but that doesn't work either.

So, I went with this minimal change that still gets us 90% there (note that before this change, it was not at all possible to flip the card back using the keyboard). WDYT?

petebacondarwin commented on aio/src/app/shared/location.service.ts:
```diff
@@ -145,11 +145,16 @@ export class LocationService {
 
     // don't navigate if external link or has extension
-    if ( anchor.href !== this.urlParser.href ||
+    if ( (!isInPageAnchor && anchor.href !== this.urlParser.href) ||
```
Why is this check needed?
(And should this fix have some associated test?)

gkalpak replied:
Since we are using a different `pathname` for in-page anchors, it is expected that `anchor.href` and `urlParser.href` will be different. For example, if we are on `https://angular.io/docs` and click a link with `href="#main-content"`, `anchor.href` will be `https://angular.io/#main-content` (incorrect), while `urlParser.href` will be `https://angular.io/docs#main-content` (correct).

If `isInPageAnchor` is true, then we know that it is an internal URL (not an external one), so we don't need to infer that by checking if `anchor.href !== urlParser.href`.

gkalpak replied:
And yes, here should be tests 😱 😱 😱

petebacondarwin replied:
I see. Thanks.

If I remember correctly, we had a workaround in the doc-gen to cope with this scenario, by adding in the current path to all local fragment URLs.

I wonder if we could remove that doc-gen fix, now that the app itself is able to cope?

gkalpak replied:
Yes, we have (a workaround in doc-gen).
And yes, we could (remove it) I think. I'll look into it.

gkalpak replied:
I looked into this and there are a couple of issues:
- The [checkAnchorLinks][1] processor in `dgeni-packages` [tries to resolve links][2] relative to the doc path and `base`, but the [resolveUrl][3] service does not resolve those correctly (i.e. is resolves `#foo` to `/#foo` instead of prepending the current doc's path).
- Our [renderLinkInfo][4] processor [cannot handle][5] URLs starting with `#`.

We could fix these, but I am not sure how much of breaking change it would be to update `resolveUrl()` in `dgeni-packages` to handle `#` links differently (even if arguably more correct 😛).

In any case, since this PR is already approved I am going to mark it for merging and we can make the changes in a follow-up PR.

[1]: https://github.com/angular/dgeni-packages/blob/4e5c3e4eee7d2bb27dd0d419bc7e211635485c3f/base/processors/checkAnchorLinks.js
[2]: https://github.com/angular/dgeni-packages/blob/4e5c3e4eee7d2bb27dd0d419bc7e211635485c3f/base/processors/checkAnchorLinks.js#L75
[3]: https://github.com/angular/dgeni-packages/blob/4e5c3e4eee7d2bb27dd0d419bc7e211635485c3f/base/services/resolveUrl.js#L13
[4]: https://github.com/angular/angular/blob/7d6ba72b90ba2aa53ee6ce8253cbc90c50e9c9ee/aio/tools/transforms/angular-base-package/processors/renderLinkInfo.js
[5]: https://github.com/angular/angular/blob/7d6ba72b90ba2aa53ee6ce8253cbc90c50e9c9ee/aio/tools/transforms/angular-base-package/processors/renderLinkInfo.js#L22

petebacondarwin commented on aio/src/app/app.component.html:
```diff
@@ -55,6 +57,7 @@
   </mat-sidenav>
 
   <main class="sidenav-content" [id]="pageId" role="main">
+    <div id="main-content" tabindex="-1"></div>
```
OOC why the -1?

gkalpak replied:
Quoting [MDN][1]:

> A negative value (usually `tabindex="-1"`) means that the element is not reachable via sequential keyboard navigation, but could be focused with JavaScript or visually by clicking with the mouse.

[1]: https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex

petebacondarwin reviewed (commented):
Nice set of updates @gkalpak - I asked a few questions so won't actually mark as approved.
But generally LGTM.

mary-poppins commented:
You can preview 07fa3e1 at https://pr43460-07fa3e1.ngbuilds.io/.

mary-poppins commented:
You can preview 6d82c69 at https://pr43460-6d82c69.ngbuilds.io/.

mary-poppins commented:
You can preview 2f3ce6f at https://pr43460-2f3ce6f.ngbuilds.io/.

jessicajaniuk reviewed (approved):
LGTM :cookie:

reviewed-for: size-tracking

dario-piotrowicz commented:
The link seems to scroll with the page... is that on purpose? :thinking:  (I was imagining it would stay fixed there)

![Peek 2021-09-18 23-01](https://user-images.githubusercontent.com/61631103/133909669-e0cdb723-bcbd-4023-90e2-45d1c0dcf809.gif)

dario-piotrowicz commented on aio/src/styles/2-modules/skip-links/_skip-links.scss:
```diff
@@ -0,0 +1,28 @@
+  margin: 5px;
+  outline: none;
+  padding: 1rem 1.5rem;
+  position: absolute;
```
```suggestion
  position: fixed;
```
(https://github.com/angular/angular/pull/43460#issuecomment-922380015)

gkalpak replied:
In all other pages I've seen "skip links" they always scroll with the page. I don't think it matters for the purpose it is intended, but I am going with the flow 😅.

dario-piotrowicz replied:
ah! you're right! :open_mouth: 

I checked the sites you've mentioned above and you are right! I guess here it feels a bit off to me because we have a fixed header so it sort of feels a bit unintentional

I thought it was just an oversight, if you're ok with it than cool :smile: 

___

(
ps: for what it's worth I noticed the youtube's one (which is not precisely to-main-content but skip-navigation, I think they are the same thing :man_shrugging:) doesn't scroll with the page! :stuck_out_tongue_winking_eye: 
![Screenshot at 2021-09-18 23-58-24](https://user-images.githubusercontent.com/61631103/133910616-c7623e71-937d-4835-b972-17fc7c92b52b.png)
)

gkalpak replied:
Fair enough :smiley:
I'll change it to be fixed. (It is true that the other pages I checked do not have a fixed header as we (and YouTube) do :wink:)

dario-piotrowicz replied:
:open_mouth: cool, thanks :smile:

dario-piotrowicz commented on aio/src/app/app.component.html:
```diff
@@ -55,6 +57,7 @@
   </mat-sidenav>
 
   <main class="sidenav-content" [id]="pageId" role="main">
+    <div id="main-content" tabindex="-1"></div>
```
is actually an extra `div` necessary? couldn't you just add the `id` to the existing `main` element?

Seems to be working the same way
![Peek 2021-09-18 23-10](https://user-images.githubusercontent.com/61631103/133909797-31aeab20-328f-4360-ab54-cec648ba34e5.gif)
:

dario-piotrowicz replied:
mh.... ah, yeah I've just noticed the `main` already has an `id` :sweat_smile: :sob: 

you could use the same `pageId` as the `href` I guess... but yeah I don't think it'd be better than how it is now :sweat_smile:

gkalpak replied:
Yeah, would do that if `main` didn't have an ID already. And I was wary of using `pageId` (which is dynamic), because I am not sure how well assistive technologies can cope with dynamically changing `href`s. (They can probably cope fine, but I wanted to err on the side of caution.)

dario-piotrowicz replied:
yeah good point! :slightly_smiling_face::+1:

dario-piotrowicz commented on aio/src/app/custom-elements/contributor/contributor-list.component.ts:
```diff
@@ -10,6 +10,7 @@ import { LocationService } from 'app/shared/location.service';
       <a *ngFor="let name of groupNames"
           class="button mat-button filter-button"
           [class.selected]="name === selectedGroup.name"
+          tabindex="0"
```
isn't `0` the default? :thinking:

dario-piotrowicz replied:
here and in the other `a`s below.... I am probably missing something :sweat_smile:

gkalpak replied:
> isn't `0` the default?

It is for interactive elements, but apparently `<a>` without an `href` are not considered interactive by some browsers (rightly so, I guess 😅). So, we need `tabindex` to indicate that this is an interactive element.

dario-piotrowicz replied:
ah.... sorry my bad, yeah, I didn't consider the `href` bit! :stuck_out_tongue_winking_eye::+1:

dario-piotrowicz replied:
or sorry..... it just came to me.... I think it would be to actually have these `a`s to `button`s instead of adding the `tabindex=0` because we sort of are trying hard to get anchors to be exactly like buttons... so why not just have them buttons right?

or may there be a particular reason why these need to be anchors? :confused: 

either way I guess that could be a bit out of scope here, but if you agree that they would be better as buttons I can look into it in a future PR if you want :slightly_smiling_face:

___

PS: I just tried locally and changing `a` to `button` is not enough as it messes up the styling a bit, but I don't think it should be too bad making it right again

gkalpak replied:
Sounds reasonable. I don't want to make more changes to this PR (esp. since it has already received some approvals), but feel free to submit a separate PR :smiley:

dario-piotrowicz replied:
awesome! will do, thanks! :smile:

jelbourn replied:
+1 to change these away from anchors in a follow-up PR. 

(same for similar anchors)

dario-piotrowicz commented on aio/src/app/custom-elements/resource/resource-list.component.html:
```diff
@@ -1,8 +1,9 @@
     <a *ngFor="let category of categories"
-      [class.selected]="category.id == selectedCategory.id"
       class="button mat-button filter-button"
+      [class.selected]="category.id == selectedCategory.id"
```
do you know if there's a reason for `==` instead of `===`?

(I was just wondering if you could update it since you're already moving it around :stuck_out_tongue_winking_eye: )

gkalpak replied:
I didn't want to make unrelated changes :grin:
There is definitely no reason (although would work the same in this case).

dario-piotrowicz reviewed (approved):
approving, for whatever it's worth it :sweat_smile: :+1:

gkalpak commented:
> approving, for whatever it's worth it 😅 👍

Thx for taking a look, @dario-piotrowicz :heart:
It is great to have another set of eyes (esp. from someone familiar with the AIO app :wink:).

gkalpak committed:
fix(docs-infra): improve accessibility of contributor cards

Improve the accessibility of contributor cards (in `/about`) by making
them easier to navigate via keyboard:

- Show the (otherwise hidden) contributor info links on focus.
- Make the "View bio" link focusable via keyboard (to give a clearer
  indication that users can click to view the bio - although technically
  the click is handled by an ancestor).
- Make the flipped cards focusable via keyboard (so it is possible to
  click them to unflip).

gkalpak committed:
fix(docs-infra): add a "skip to content" link for better accessibility

Add a link at the top of the page that allows skipping directly to the
main content area. This allows people using assistive technologies to
skip navigation-related parts of the page which are presented before the
main content.

Related resources:
- [www.w3.org](https://www.w3.org/TR/WCAG20-TECHS/G1.html)
- [accessibility.oit.ncsu.edu](https://accessibility.oit.ncsu.edu/it-accessibility-at-nc-state/developers/accessibility-handbook/mouse-and-keyboard-events/skip-to-main-content)
- [css-tricks.com](https://css-tricks.com/how-to-create-a-skip-to-content-link)

---
Example implementations on other websites:

- **[accessibility.oit.ncsu.edu](https://accessibility.oit.ncsu.edu/)**:
  ![accessibility.oit.ncsu.edu](https://user-images.githubusercontent.com/8604205/133285834-db1314e6-ff71-4862-8f80-c8a9f69083d9.png)

- **[smile.amazon.com](https://smile.amazon.com/)**:
  ![smile.amazon.com](https://user-images.githubusercontent.com/8604205/133285854-1163e733-8de3-4244-bc23-5cd5a943001d.png)

- **[chase.com](https://chase.com/)**:
  ![chase.com](https://user-images.githubusercontent.com/8604205/133293396-ade9d95f-56ce-4cdc-8926-b4cd2dc6c3a5.png)

- **[css-tricks.com](https://css-tricks.com/)**:
  ![css-tricks.com](https://user-images.githubusercontent.com/8604205/133285845-702045ee-088a-40f1-bb26-21adebb02505.png)

- **[github.com/angular/angular](https://github.com/angular/angular)**:
  ![github.com/angular/angular](https://user-images.githubusercontent.com/8604205/133921259-9fd759b9-f862-47c3-b069-643912df17a1.png)

- **[news.sky.com](https://news.sky.com/)**:
  ![news.sky.com](https://user-images.githubusercontent.com/8604205/133285863-89946096-5353-4b2b-ab69-eb748bdcfe06.png)

- **[youtube.com](https://youtube.com/)**:
  ![youtube.com](https://user-images.githubusercontent.com/8604205/133921261-a68a12c1-1aea-4ad0-8457-b4eaf8ed3497.png)

---
Angular.io implementation:

- **[angular.io](https://angular.io/)**:
  ![angular.io](https://user-images.githubusercontent.com/8604205/133409634-0054625c-d14b-4ebe-bb4c-142cfc524d98.png)

Fixes #42936

gkalpak committed:
fix(docs-infra): improve accessibility of button groups

Improve the accessibility of button groups (such as those seen in
`/about` and `/resources` to choose between the different categories)
by making them easier to navigate via keyboard:

- Make the individual buttons focusable via keyboard.
  (Previously, they were not focusable - despite being anchor elements,
  because they were lacking an `href` attribute.)

- Give focused buttons a different style.

gkalpak committed:
fix(docs-infra): correctly handle in-page anchor links

Previously, anchor links that were supposed to point to fragments inside
the current page (e.g. `href="#something"`) would end up navigating to
the homepage (`/`). This was due to them being resolved relative to the
base URL, which is set to `/` (via `<base href="/">`). See also
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base#in-page_anchors.

This commit fixes this issue by handling such links specially in
`LocationService#handleAnchorClick()` and prepending the current path to
the URL.

This will also be useful for implementing a "skip to content" link as
described in #42936.

gkalpak committed:
fix(docs-infra): focus elements when scrolling to them (if focusable)

When using `ScrollService#scrollToElement()` to scroll to a specific
element, also focus the element if it is focusable (i.e. if it has a
`focus()` method).

This will be useful for implementing a "skip to content" link as
described in #42936.

gkalpak committed:
fixup! fix(docs-infra): correctly handle in-page anchor links

gkalpak committed:
fixup! fix(docs-infra): improve accessibility of contributor cards

gkalpak committed:
fixup! fix(docs-infra): add a "skip to content" link for better accessibility

mary-poppins commented:
You can preview 586c683 at https://pr43460-586c683.ngbuilds.io/.

petebacondarwin reviewed (approved):
LGTM both after my previous comments, and for size-tracking!

Reviewed-for: size-tracking

jelbourn commented on aio/src/styles/2-modules/skip-links/_skip-links.scss:
```diff
@@ -0,0 +1,28 @@
+
+.skip-to-content-link {
+  @include mat.elevation(6);
+  background: map.get(mat.$grey-palette, 50);
```
Does these colors in the rule need to change for light/dark mode?

gkalpak replied:
Good question 🤔
It is always shown on top of our blue toolbar, so I thought it is fine to always show with a white background.

Do you think a dark background would be better in dark mode?

jelbourn replied:
I'm satisfied so long as the text is readable and it doesn't look super out of place in either mode

jelbourn reviewed (approved):
lgtm
